sematic/resolvers/state_machine_resolver.py

```diff
@@ -23,50 +23,58 @@ def __init__(self):
 
     @property
     def _root_future(self) -> AbstractFuture:
+        if len(self._futures) == 0:
+            raise RuntimeError("No root future: no futures were enqueued.")
+
         return self._futures[0]
 
     def resolve(self, future: AbstractFuture) -> typing.Any:
         with self._catch_resolution_errors():
-            self._enqueue_root_future(future)
-
-            self._register_signal_handlers()
-
-            logger.info(f"Starting resolution {future.id}")
-
-            self._resolution_will_start()
-
-            while not future.state.is_terminal():
-                for future_ in self._futures:
-                    if future_.state == FutureState.CREATED:
-                        self._schedule_future_if_args_resolved(future_)
-                        continue
-                    if future_.state == FutureState.RETRYING:
-                        self._execute_future(future_)
-                        continue
-                    if future_.state == FutureState.RAN:
-                        self._resolve_nested_future(future_)
-                        continue
-
-                    # should be unreachable code, here for a sanity check
-                    if (
-                        future_.state != FutureState.SCHEDULED
-                        and not future_.state.is_terminal()
-                    ):
-                        raise RuntimeError(
-                            f"Illegal state: future {future_.id} in state {future_.state}"
-                            " when it should have been already processed"
-                        )
-
-                self._wait_for_scheduled_runs()
-
-            if future.state == FutureState.RESOLVED:
-                self._resolution_did_succeed()
-            elif future.state == FutureState.CANCELED:
-                return
-            else:
-                raise RuntimeError("Unresolved Future after resolver call.")
-
-            return future.value
+            self._seed_graph(future)
+            self._resolution_loop()
+
+        return self._root_future.value
+
+    def _seed_graph(self, future):
```

Comment:
Done.

Reply:
I was thinking more in terms of documenting it for the generic purpose, since it is overridden. Something along the lines of
```
"""Set the initial futures for the resolution

This is a stub that can be overridden by child classes.
The resolver will evolve the DAG using this initial set of futures.
The root future should always be element 0 of `self._futures`,
there are no requirements on order of other futures. The default
implementation just seeds with the root future itself and no others.
"""
```

Reply:
Done.